feat(ui): add previous runs section to sensor details page

puretension opened the pull request:
<!-- markdownlint-disable MD041 -- this is rendered within existing HTML, so allow starting without an H1 -->

<!--

### Before you open your PR

- Run `make pre-commit -B` to fix codegen and lint problems (build will fail).
- [Signed-off your commits](https://github.com/apps/dco/) (otherwise the DCO check will fail).
- Used [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/).

### When you open your PR

- PR title format should also conform to [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
- Create the PR as draft.
- Once builds are green, mark your PR "Ready for review".

When changes are requested, please address them and then dismiss the review to get it reviewed again.

-->

<!-- Does this PR fix an issue -->

Fixes #14809

### Motivation

Sensor details page lacks Previous Runs visibility, making it difficult for users to locate and monitor workflows triggered by sensor events. This creates inconsistent UX compared to CronWorkflow which provides clear Previous Runs section.

### Modifications

- Added Previous Runs section below Sensor editor tabs using `workflows.argoproj.io/sensor` label filtering
- Implemented identical UI pattern as CronWorkflow with WorkflowDetailsList component for consistency
- Fixed empty state handling with proper array length check to display triggered workflows correctly

https://github.com/user-attachments/assets/0230c8b0-0784-466a-9d4e-d810a16f8783

### Verification

- Sensor details page now displays Previous Runs section with triggered workflows
- Workflows are properly filtered by `workflows.argoproj.io/sensor=sensor-name` label
- Empty state shows appropriate message when no workflows exist
- UI matches CronWorkflow Previous Runs behavior for consistent user experience

### Documentation

<!-- TODO: Say how you have updated the documentation or explain why this isn't needed here -->
<!-- Required for features: Explain how the user will discover this feature through documentation and examples -->

No documentation changes needed as this follows existing CronWorkflow Previous Runs pattern that users are already familiar with.

<!--
### Beyond this PR

Thank you for submitting this! Have you ever thought of becoming a Reviewer or Approver on the project?

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable.
See [Sustainability Effort](https://github.com/argoproj/argo-workflows/blob/main/community/sustainability_effort.md) for more information.

-->

puretension committed:
feat(ui): add label filter sync with URL query params

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
feat(ui): add feature description for label filter URL sync

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
feat(ui): add sensor label to workflows model

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
feat(ui): add previous runs section to sensor details page

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
feat(ui): add feature description for previous runs section to sensor details page

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
feat(ui): add feature description for previous runs section to sensor details page

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
feat(ui): add feature description for previous runs section to sensor details page

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

Joibel commented on .features/pending/14807-workflow-templates-label-filter-sync.md:
```diff
@@ -0,0 +1,9 @@
+Component: UI
```
This file shouldn't be in this PR.

puretension replied:
@Joibel  Fixed! I removed the incorrect file from the PR. It was accidentally included from a wrong branch merge.
(I think it came my previous work - https://github.com/argoproj/argo-workflows/pull/14816)

puretension committed:
feat(ui): delete incorrect file from PR

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

Joibel commented:
> Hi, thanks for working on it. Any update?

Copilot had some suggestions, can you look at those please or tell me they're wrong?

tico24 commented on .features/pending/14809-sensor-previous-runs.md:
```diff
@@ -0,0 +1,8 @@
+Component: UI
+Issues: 14809
+Description: Add Previous Runs section to Sensor details page to display workflows triggered by sensors
```
The body underneath (lines 6 to 8) is what ends up in front of users, and the template asks it to explain when you'd want the feature, in present tense, written to the reader. Something like:

```markdown
The Sensor details page now lists the Workflows that this Sensor's triggers created, below the Sensor editor.
Use it to check whether a Sensor is firing and to open the Workflows it produced, instead of going to the workflow list and filtering by label yourself.
When the Sensor has not triggered anything, the page tells you so.
```

One sentence per line, as you already have.

puretension replied:
Adopted your wording, thanks. The body is now:

```
The Sensor details page now lists the Workflows that this Sensor's triggers created, below the Sensor editor.
Use it to check whether a Sensor is firing and to open the Workflows it produced, instead of going to the workflow list and filtering by label yourself.
When the Sensor has not triggered anything, the page tells you so.
The list works the same way as the one on the CronWorkflow details page.
```

tico24 commented on .features/pending/14809-sensor-previous-runs.md:
```diff
@@ -0,0 +1,8 @@
+Description: Add Previous Runs section to Sensor details page to display workflows triggered by sensors
+Author: [puretension](https://github.com/puretension)
+
+- Added Previous Runs section below Sensor editor tabs using `workflows.argoproj.io/sensor` label filtering
```
I can't find anything that sets `workflows.argoproj.io/sensor`. This PR adds the constant to `ui/src/shared/models/workflows.ts`, but nothing in this repo writes the label, and neither does Argo Events.

What Argo Events puts on a Workflow created by a Sensor trigger is `events.argoproj.io/sensor`, in both `pkg/sensors/triggers/standard-k8s/standard-k8s.go` and `pkg/sensors/triggers/argo-workflow/argo-workflow.go`. This UI already reads that label in `ui/src/event-flow/build-graph.ts` and filters on it in `event-flow-page.tsx`.

If that holds, the section shows "No workflows have been triggered by this sensor yet" for a Sensor that's been firing all day. Worth confirming against a cluster before anything else:

```bash
kubectl get workflows -l events.argoproj.io/sensor=<sensor-name>
kubectl get workflows -l workflows.argoproj.io/sensor=<sensor-name>
```

Which label the code should use is for the approvers to decide. The note just shouldn't describe a filter that matches nothing.

puretension replied:
You were right, and I've confirmed it on a cluster rather than just by reading. Your two commands, against a Sensor that had been firing for a few minutes:

```console
$ kubectl get workflows -n argo -l events.argoproj.io/sensor=calendar-sensor
NAME                       STATUS      AGE
calendar-triggered-d45cz   Succeeded   3m32s
calendar-triggered-9qsvl   Succeeded   3m47s
calendar-triggered-84g4f   Succeeded   4m1s
... (10 total)

$ kubectl get workflows -n argo -l workflows.argoproj.io/sensor=calendar-sensor
No resources found in argo namespace.
```

The full label set Argo Events puts on a triggered Workflow:

```
events.argoproj.io/action-timestamp = 1787016448030
events.argoproj.io/sensor           = calendar-sensor
events.argoproj.io/trigger          = calendar-workflow-trigger
workflows.argoproj.io/completed     = false
workflows.argoproj.io/creator       = system-serviceaccount-argo-operate-workflow-sa
workflows.argoproj.io/phase         = Running
```

Matching the source you pointed at, in argo-events v1.9.11: `pkg/sensors/triggers/argo-workflow/argo-workflow.go:115` sets `submittedWFLabels["events.argoproj.io/sensor"] = t.Sensor.Name`, and `pkg/sensors/triggers/standard-k8s/standard-k8s.go` sets the same key on Create.

Fixed — the query uses `events.argoproj.io/sensor`. I also moved the constant: it's an Argo Events label, not a `workflows.argoproj.io/*` one, so keeping it in the `labels` map in `workflows.ts` was misleading. It now lives in `ui/src/shared/models/events.ts`:

```ts
// labels that Argo Events puts on the resources its Sensor triggers create
export const eventsLabels = {
    sensor: 'events.argoproj.io/sensor'
};
```

and the invented `workflows.argoproj.io/sensor` constant is gone.

For the record on how this got through: the recording in the original description was made against hand-written Workflow YAMLs that I had labelled `workflows.argoproj.io/sensor` myself, so the section looked populated. It was never tested against a Sensor that actually fired. That's on me.

tico24 commented on .features/pending/14809-sensor-previous-runs.md:
```diff
@@ -0,0 +1,8 @@
+Author: [puretension](https://github.com/puretension)
+
+- Added Previous Runs section below Sensor editor tabs using `workflows.argoproj.io/sensor` label filtering
+- Implemented identical UI pattern as CronWorkflow with WorkflowDetailsList component for consistency
```
`WorkflowDetailsList` is an internal component name, and "identical UI pattern as CronWorkflow" describes how you built it rather than what someone gets. Whoever reads this in the release notes has never seen the diff.

If the resemblance to CronWorkflow is worth keeping, put it in terms of the page: "The list works the same way as the one on the CronWorkflow details page." Otherwise I'd drop the line.

puretension replied:
I dropped the component name and the "identical UI pattern" phrasing, and used your sentence instead: "The list works the same way as the one on the CronWorkflow details page."

tico24 commented on .features/pending/14809-sensor-previous-runs.md:
```diff
@@ -0,0 +1,8 @@
+
+- Added Previous Runs section below Sensor editor tabs using `workflows.argoproj.io/sensor` label filtering
+- Implemented identical UI pattern as CronWorkflow with WorkflowDetailsList component for consistency
+- Fixed empty state handling with proper array length check to display triggered workflows correctly
```
This describes fixing code that the same PR adds, so from the outside there was never a broken empty state. Someone reading the release notes would go looking for the version that had the bug.

I'd drop it and instead say what a user sees when the Sensor hasn't triggered anything, which the rewrite on line 3 covers.

puretension replied:
Yes I agreed and dropped. there was never a released version with that bug, so nobody should go looking for it. What a user sees when the Sensor hasn't triggered anything is covered by the third line of the rewritten body.

puretension commented on ui/src/sensors/sensor-details.tsx:
```diff
@@ -142,6 +157,15 @@ export function SensorDetails({match, location, history}: RouteComponentProps<an
                     />
                 )}
+                <>
+                    {!workflows || workflows.length === 0 ? (
```
I tried this and it crashes the page, so I've fixed it a different way.

`services.workflows.list()` hands back the argo-server response as-is, and when nothing matches the selector the server sends `items: null`, not `[]`:

```console
$ curl -sk ".../api/v1/workflows/argo?listOptions.labelSelector=events.argoproj.io%2Fsensor%3Dnever-fired-sensor"
{"metadata":{...},"items":null}
```

So `setWorkflows(workflowList.items)` puts `null` into state, and `workflows.length` throws. With the suggested change applied, a Sensor that has never triggered anything renders the error boundary instead of the empty state:

```
TypeError: Cannot read properties of null (reading 'length')
    at SensorDetails (src/sensors/sensor-details.tsx:189)
```

That's also why the check in `cron-workflow-details.tsx` is `!workflows` — `null` is what actually arrives there on an empty list, so that guard is load-bearing, not redundant.

Rather than keep the double check, I normalise at the boundary, which keeps your simplification at the render site:

```tsx
setWorkflows(workflowList.items || []);
...
{workflows.length === 0 ? (
    <ZeroState title='No previous runs'>...</ZeroState>
) : (
    <WorkflowDetailsList workflows={workflows} columns={columns} />
)}
```

Verified both ways on a live cluster: a Sensor with 10 triggered Workflows lists them, and a Sensor that has never fired shows "No previous runs".

puretension commented on ui/src/sensors/sensor-details.tsx:
```diff
@@ -66,6 +71,16 @@ export function SensorDetails({match, location, history}: RouteComponentProps<an
+            const workflowsInfo = await services.info.getInfo();
+
+            setWorkflows(workflowList.items);
+            setColumns(workflowsInfo.columns);
```
Done:

```tsx
useEffect(() => {
    (async () => {
        const workflowList = await services.workflows.list(namespace, null, [`${models.eventsLabels.sensor}=${name}`], {limit: 50});
        const workflowsInfo = await services.info.getInfo();

        setWorkflows(workflowList.items || []);
        setColumns(workflowsInfo.columns);
    })().catch(setError);
}, [namespace, name]);
```

`<ErrorNotice error={error} />` is already rendered at the top of the page, so a failed list call now surfaces there instead of being swallowed.

puretension commented:
Rebase artefact, dropped. It came from #14816, merged in October as `c13a492`
this branch was still carrying that PR's pre-merge copy of the file.

Worse than just being unrelated: the stale copy would have reverted two later fixes to the same file — #15796 (populate URL filter params on first load) and the `target='_blank'` change in #15860. 
That's what the merge conflict was.

The branch is now rebased onto current `main`, so the conflict is resolved and the PR is three files:

- `ui/src/sensors/sensor-details.tsx`
- `ui/src/shared/models/events.ts`
- `.features/pending/14809-sensor-previous-runs.md`

puretension commented on ui/src/sensors/sensor-details.tsx:
```diff
@@ -142,6 +157,13 @@ export function SensorDetails({match, location, history}: RouteComponentProps<an
+                    </ZeroState>
+                ) : (
+                    <WorkflowDetailsList workflows={workflows} columns={columns} />
+                )}
```
Both parts reproduce, so I've fixed it. Thanks.

The flash, sampling the page every 100ms on a Sensor with 10 triggered Workflows:

```
ZeroState first seen: 500ms
list first seen:      900ms   -> "No previous runs" visible for ~400ms
```

The failure case is the worse of the two. Dropping the connection to argo-server and loading the page gave:

```
 Gateway Timeout
No previous runs
No workflows have been triggered by this sensor yet.
```

So the page asserted the Sensor had never triggered anything when the query had simply failed — the `.catch(setError)` added earlier in this review sets the error but leaves `workflows` at `[]`, which keeps the empty-state branch true.

Fixed by making "not loaded yet" a distinct state rather than overloading `[]`:

```tsx
const [workflows, setWorkflows] = useState<Workflow[]>();   // undefined until the query resolves
...
setWorkflows(workflowList.items || []);                     // argo-server sends items: null when empty
...
{workflows &&
    (workflows.length === 0 ? (
        <ZeroState title='No previous runs'>...</ZeroState>
    ) : (
        <WorkflowDetailsList workflows={workflows} columns={columns} />
    ))}
```

`undefined` while loading or after a failure renders nothing below the editor, and `<ErrorNotice>` at the top of the page carries the failure on its own.

Re-checked all three states on a live cluster (kind, argo-workflows v4.1.1, argo-events v1.9.11):

| state | before | after |
| --- | --- | --- |
| 10 triggered Workflows | ZeroState for ~400ms, then the list | list at 300ms, ZeroState never appears |
| Sensor that never fired | "No previous runs" | "No previous runs" |
| argo-server unreachable | error **and** "No previous runs" | error only |

No runtime errors in the console in any of the three.
